### Notebook 1: Data Loading and Cleaning

Loading and cleaning the data upon first look.

In this notebook, I will first define the purpose of this project where I discuss the reasons for doing my analysis. I then will begin the steps to loading and cleaning the data


______________________

#### __Notebook Contents__

__1.1__ Introduction

__1.1.1__ Workflow Diagram

__1.2__ Loading Cleaning & Checking Data

__1.3__ Investigating Columns

__1.4__ Cleaning Text in Review Column

__1.5__ Addressing Null Values

__1.6__ Checking for Duplicates

__1.7__ Correcting Data Types

__1.8__ Feature Creation

__1.9__ Saving Work


#### __1.1 Introduction__

**Problem Statement**

As of July 2023, the global size of the airline industry is estimated at a whopping $841.5bn, a 7% increase from the year before. Although one could argue that the industry is very stable due to the sheer demand from consumers, airlines could still use with improving customer satisfaction with the aim of becoming the favoured option in a highly saturated market.

The problem at hand is to develop a data-driven solution to improve the quality of airline services and ultimately earn prospective airlines a good reputation in relation to their competitiors by analysing and extracting valuable insights from passenger reviews. Airlines receive an immeasurable volume of feedback from customers through various online platforms such as review websites and customer surveys. These reviews encompass a wide range of sentiments, including positive and negative feedback, some a bit too difficult to decipher. **We aim to produce an automated solution to decipher sentiment**.

As mentioned earlier, this industry's stability is almost as similar to ones of other indutries such as commodities; so why would people leave reviews? good or bad? especially if they are more than likely to reuse these services, better yet - what experiences wow passengers so much that they would recommend the airline?

___________________________________________________________________________________________________________________________________________________________________________________________________________________________

The goal of this project is to explore the main pull factors that make a passenger's airline experience so great that not only would they leave a positive review but recommend the services to other curious researchers. I hope the conclusions drawn by the end of this project provides useful and actionable insights for airline managers on what levels of services are to be provided to receive a high rating score and benefit from word of mouth. This would especially help attract some new passengers, perhaps be in a better position competitvely, with the ultimate aim of being the most trusted and favoured airline.

**Data Collection**

The data used in this project was originally web-scraped from the Skytrax website, but later published on Kaggle by user `Efehan` titled, `Skytrax Airline Reviews`, and contains a large list of airline review scores along with the following attributes: airline, overall, author, review_date, customer_review, aircraft, traveller_type, cabin, date_flown, seat_comfort, cabin_service, food_bev, entertainment, ground_service, value_for_money, recommended. This dataset contains a mixture of both numeric and non-numeric data types.

A direct download link to the dataset can be found [here](https://drive.google.com/uc?export=download&id=1ht8AOCyKsbBOb5ys58x5EU1f1do0yGRB):

#### __1.1.1 Workflow__

![Alt text](image.png)

**Importing Libraries**

These are the libraries required to load, clean and prepare the data for analysis.

In [1]:
#Standard Libraries

import numpy as np
import pandas as pd
import joblib
import re


#### __1.2 Loading, Cleaning & Checking Data__

#### Data and Column Descriptions

Before analysing the data, I want to take a quick look and explore the data and deal with any cleaning related issues accordingly.

In [2]:
#Read in data from my data folder
airlinedf = pd.read_excel('/Users/faisal/BS_Capstone/capstone-project-fai22399-git/data/airline_reviews.xlsx', index_col=None, header=0)


In [3]:
#check if data has been read in by viewing the first 5 rows
airlinedf.head()

airline  overall               author   review_date  \
0               NaN      NaN                  NaN           NaN   
1  Turkish Airlines      7.0  Christopher Hackley  8th May 2019   
2               NaN      NaN                  NaN           NaN   
3  Turkish Airlines      2.0        Adriana Pisoi  7th May 2019   
4               NaN      NaN                  NaN           NaN   

                                     customer_review aircraft  traveller_type  \
0                                                NaN      NaN             NaN   
1  ✅ Trip Verified | London to Izmir via Istanb...      NaN        Business   
2                                                NaN      NaN             NaN   
3  ✅ Trip Verified | Istanbul to Bucharest. We ...      NaN  Family Leisure   
4                                                NaN      NaN             NaN   

           cabin                         route           date_flown  \
0            NaN                           NaN                  NaN   
1  Economy Class  London to Izmir via Istanbul  2019-05-01 00:00:00   
2            NaN                           NaN                  NaN   
3  Economy Class         Istanbul to Bucharest  2019-05-01 00:00:00   
4            NaN                           NaN                  NaN   

   seat_comfort  cabin_service  food_bev  entertainment  ground_service  \
0           NaN            NaN       NaN            NaN             NaN   
1           4.0            5.0       4.0            4.0             2.0   
2           NaN            NaN       NaN            NaN             NaN   
3           4.0            1.0       1.0            1.0             1.0   
4           NaN            NaN       NaN            NaN             NaN   

   value_for_money recommended  
0              NaN         NaN  
1              4.0         yes  
2              NaN         NaN  
3              1.0          no  
4              NaN         NaN

Immediately, I can notice a pattern whereby every other row is filled with completely `NaN` values. I will look further into the dataset to confirm this early finding and then deal with it accordingly, perhaps by dropping null rows.

In [4]:
#Returns the shape of the DataFrame
print(f"Shape of DataFrame: {airlinedf.shape}")

Shape of DataFrame: (131895, 17)


In [5]:
print("List of Columns")
list(airlinedf.columns)

List of Columns


['airline',
 'overall',
 'author',
 'review_date',
 'customer_review',
 'aircraft',
 'traveller_type',
 'cabin',
 'route',
 'date_flown',
 'seat_comfort',
 'cabin_service',
 'food_bev',
 'entertainment',
 'ground_service',
 'value_for_money',
 'recommended']

**Column Description:**

`airline`: Name of the airline.

`overall`: Overall point given to the trip between 1 to 10.

`author`: Author of the trip(Passenger Name)

`review_date`: Date of the Review

`customer_review`: Review of the customers in free text format

`aircraft`: Type of the aircraft

`traveller_type`: Type of traveller (e.g. business, leisure)

`cabin`: Cabin at the flight

`date_flown`: Flight date

__The following are categories that link to the overall satisfaction__:

`seat_comfort`: Rated between 1-5

`cabin_service`: Rated between 1-5

`food_bev`: Rated between 1-5

`entertainment`: Rated between 1-5

`ground_service`: Rated between 1-5

`value_for_money`: Rated between 1-5

`recommended`: Binary, target variable. (yes,no)

In [6]:
airlinedf.describe()

overall  seat_comfort  cabin_service      food_bev  entertainment  \
count  64017.000000  60681.000000   60715.000000  52608.000000   44193.000000   
mean       5.145430      2.952160       3.191814      2.908170       2.863372   
std        3.477532      1.441362       1.565789      1.481893       1.507262   
min        1.000000      1.000000       1.000000      1.000000       1.000000   
25%        1.000000      1.000000       2.000000      1.000000       1.000000   
50%        5.000000      3.000000       3.000000      3.000000       3.000000   
75%        9.000000      4.000000       5.000000      4.000000       4.000000   
max       10.000000      5.000000       5.000000      5.000000       5.000000   

       ground_service  value_for_money  
count    39358.000000     63975.000000  
mean         2.692820         2.943962  
std          1.612215         1.587370  
min          1.000000         1.000000  
25%          1.000000         1.000000  
50%          3.000000         3.000000  
75%          4.000000         4.000000  
max          5.000000         5.000000

**Overview:**

We run a .describe() function to understand the numerical values within the dataset

**Key Points**
- The range of ratings for specific categories is from 1 to 5
- The range of ratings for the overall is from 1 to 10
- The average overall rating is 5.14, which naturally would be the case. 

________

#### __1.3 Investigating Columns__

In [7]:
#Get an overview of the dataset, how many null values are there?
airlinedf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 131895 entries, 0 to 131894
Data columns (total 17 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   airline          65947 non-null  object 
 1   overall          64017 non-null  float64
 2   author           65947 non-null  object 
 3   review_date      65947 non-null  object 
 4   customer_review  65947 non-null  object 
 5   aircraft         19718 non-null  object 
 6   traveller_type   39755 non-null  object 
 7   cabin            63303 non-null  object 
 8   route            39726 non-null  object 
 9   date_flown       39633 non-null  object 
 10  seat_comfort     60681 non-null  float64
 11  cabin_service    60715 non-null  float64
 12  food_bev         52608 non-null  float64
 13  entertainment    44193 non-null  float64
 14  ground_service   39358 non-null  float64
 15  value_for_money  63975 non-null  float64
 16  recommended      64440 non-null  object 
dtypes: float64

In [8]:
#Count of null values
airlinedf.isna().sum()

airline             65948
overall             67878
author              65948
review_date         65948
customer_review     65948
aircraft           112177
traveller_type      92140
cabin               68592
route               92169
date_flown          92262
seat_comfort        71214
cabin_service       71180
food_bev            79287
entertainment       87702
ground_service      92537
value_for_money     67920
recommended         67455
dtype: int64

Interesting discovery here... as mentioned earlier, it appeared that every other collumn was filled with `null` values. 

Now, I can see that the null values are almost identical to the non-null values from the `.info()` code block above. I can now be certain that there are double rows and can therefore drop the rows where all datapoints are null.

In [9]:
#drop all rows where all data points are null using the how='all' condition
cleandf = airlinedf.dropna(axis=0, how='all').copy().reset_index(drop=True).copy()

In [10]:
#Check the shape of the DataFrame
print(f"Shape of DataFrame: {cleandf.shape}")

Shape of DataFrame: (65947, 17)


I can now see that the shape of the DataFrame has reduced massively from 131,895 rows to just 65,947 rows, which seems to correlate perfectly with the data in the `airline` and `customer_review` columns.

This dataset is close to being cleaned and ready for analysis, however there are a few columns which still look like they are filled with too many null values for our likening. I will proceed to handle them accordingly.

In [11]:
#re-load dataframe to view changes
cleandf.head()

airline  overall               author   review_date  \
0  Turkish Airlines      7.0  Christopher Hackley  8th May 2019   
1  Turkish Airlines      2.0        Adriana Pisoi  7th May 2019   
2  Turkish Airlines      3.0            M Galerko  7th May 2019   
3  Turkish Airlines     10.0          Zeshan Shah  6th May 2019   
4  Turkish Airlines      1.0           Pooja Jain  6th May 2019   

                                     customer_review aircraft  traveller_type  \
0  ✅ Trip Verified | London to Izmir via Istanb...      NaN        Business   
1  ✅ Trip Verified | Istanbul to Bucharest. We ...      NaN  Family Leisure   
2  ✅ Trip Verified | Rome to Prishtina via Ista...      NaN        Business   
3  ✅ Trip Verified | Flew on Turkish Airlines I...     A330    Solo Leisure   
4  ✅ Trip Verified | Mumbai to Dublin via Istan...      NaN    Solo Leisure   

           cabin                           route           date_flown  \
0  Economy Class    London to Izmir via Istanbul  2019-05-01 00:00:00   
1  Economy Class           Istanbul to Bucharest  2019-05-01 00:00:00   
2  Economy Class  Rome to Prishtina via Istanbul  2019-05-01 00:00:00   
3  Economy Class    Washington Dulles to Karachi           April 2019   
4  Economy Class   Mumbai to Dublin via Istanbul  2019-05-01 00:00:00   

   seat_comfort  cabin_service  food_bev  entertainment  ground_service  \
0           4.0            5.0       4.0            4.0             2.0   
1           4.0            1.0       1.0            1.0             1.0   
2           1.0            4.0       1.0            3.0             1.0   
3           4.0            5.0       5.0            5.0             5.0   
4           1.0            1.0       1.0            1.0             1.0   

   value_for_money recommended  
0              4.0         yes  
1              1.0          no  
2              2.0          no  
3              5.0         yes  
4              1.0          no

#### __1.4 Cleaning Text in Review Column__

The next observation of the data is that there are some odd characters at the beginning of each review, some sort of symbol perhaps an emoji, we will proceed to clean the text as we will end up with that characterised symbol as a highly impactive 'word' when it comes to our sentiment analysis.

Another thing to mention is that at the begining of each review, the author has inputed their location of origin as well as their destination. In this specific case, this information will not be useful as at the point of modelling, it may just get an output saying the location is the highest positive/negative sentiment i.e. `london` as opposed to the real push and pull factors we are aiming to identify i.e. problems/praises regarding customer service. We will aim to sort this out during our preprocessing stage before modelling.

Upon further inspection, we notice that a large number of reviews have a prefix along the lines of `verified review`, we would want to remove these as part of the text cleaning to ensure no bias in the data. Given the fact that this prefix would appear in majority of reviews, at the stage of modelling, this would appear as an influential indicator of our sentiment analysis. In our use case, this provides no context whatsoever to the inference problem.

In [12]:
# #clean text
def cleanedtext(text):
    text = re.sub(r'[^\w\s]', ' ', text)
    text = text.lower()
    text = text.lstrip("✠verified review").lstrip("verified review").lstrip("✠trip").lstrip('verified').lstrip(" trip verified").lstrip("... verified review")
    return text

cleandf['customer_review'] = cleandf['customer_review'].apply(cleanedtext)

In [13]:
#Check function output
cleandf['customer_review'].sample(5)

27048    have not flown china southern for many years d...
20879    seoul gimpo to jeju  the flight that i booked ...
56945    booking was easy and transparent  paid 168â   ...
54785    lauderdale to baltimore with spirit airlines  ...
2083     stockholm to doha with qatar airways  sas grou...
Name: customer_review, dtype: object

This may not be the cleanest of data with regards to grammatical errors, however, it is suitable enough to proceed. 


________

#### __1.5 Addressing Null Values__

In [14]:
#check the percentage of null values
print(f'Percentage of null columns')
round(cleandf.isna().sum()/len(cleandf)*100, 2)

Percentage of null columns


airline             0.00
overall             2.93
author              0.00
review_date         0.00
customer_review     0.00
aircraft           70.10
traveller_type     39.72
cabin               4.01
route              39.76
date_flown         39.90
seat_comfort        7.99
cabin_service       7.93
food_bev           20.23
entertainment      32.99
ground_service     40.32
value_for_money     2.99
recommended         2.29
dtype: float64

Immediately the column drawn to my attention is the `aircraft` column. 

With 70% of data being null, this is too significant to my dataset and luckily, due to the fact of this data not being relevant to my final model, I have full confidence to drop this entire column, but make sure to copy as good practise.

In [15]:
#dropping aircraft column
cleandf = cleandf.drop('aircraft', axis=1).copy()

In [16]:
#check the percentage of null values (Iteration 1)
print(f'Percentage of null columns')
round(cleandf.isna().sum()/len(cleandf)*100, 2)

Percentage of null columns


airline             0.00
overall             2.93
author              0.00
review_date         0.00
customer_review     0.00
traveller_type     39.72
cabin               4.01
route              39.76
date_flown         39.90
seat_comfort        7.99
cabin_service       7.93
food_bev           20.23
entertainment      32.99
ground_service     40.32
value_for_money     2.99
recommended         2.29
dtype: float64

This now looks a bit cleaner, but we cannot ignore the other columns which have over 30% null values and also not significant to the model.

`traveller_type`: Although this may help differentiate between the type of customer, seeing as the main focus of our data is based on text, this is something I will not look into.

`route`: Although one could argue, the route of the flight could impact the overall experience, the airline potentially has little no effect on the route and therefore would be classed as an external factor. This would not be useful in the model.

`date_flown`: This date most likely corresponds with the data within the `review_date` column and would therefore be classified as a duplicate column, and for that reason, I can drop this column.

In [17]:
#drop the columns not useful to the model, with a high number of null values
cleandf = cleandf.drop(columns=['route','date_flown', 'traveller_type'], axis=1).copy()

In [18]:
#check the percentage of null values (Iteration 2)
print(f'Percentage of null columns')
round(cleandf.isna().sum()/len(cleandf)*100, 2)

Percentage of null columns


airline             0.00
overall             2.93
author              0.00
review_date         0.00
customer_review     0.00
cabin               4.01
seat_comfort        7.99
cabin_service       7.93
food_bev           20.23
entertainment      32.99
ground_service     40.32
value_for_money     2.99
recommended         2.29
dtype: float64

Now, I am left with still quite a few nulls, but the majority fall within columns which directly impact our target column of `recommended`. 

Due to these columns being based on rating but for respective categories, the most suitable solution would be to fill in the null values with the mean of each column. This will ensure a fair balance in the distribution of data (to be shown in the EDA notebook).

In [19]:
# filling null values with the mean of the column
cleandf['ground_service']= cleandf['ground_service'].fillna(cleandf['ground_service'].mean())
cleandf['seat_comfort']= cleandf['seat_comfort'].fillna(cleandf['seat_comfort'].mean())
cleandf['cabin_service']= cleandf['cabin_service'].fillna(cleandf['cabin_service'].mean())
cleandf['food_bev']= cleandf['food_bev'].fillna(cleandf['food_bev'].mean())
cleandf['entertainment']= cleandf['entertainment'].fillna(cleandf['entertainment'].mean())
cleandf['value_for_money']= cleandf['value_for_money'].fillna(cleandf['value_for_money'].mean())

In [20]:
#check the percentage of null values (Iteration 3)
print(f'Percentage of null columns')
round(cleandf.isna().sum()/len(cleandf)*100, 2)

Percentage of null columns


airline            0.00
overall            2.93
author             0.00
review_date        0.00
customer_review    0.00
cabin              4.01
seat_comfort       0.00
cabin_service      0.00
food_bev           0.00
entertainment      0.00
ground_service     0.00
value_for_money    0.00
recommended        2.29
dtype: float64

This looks much cleaner! There are still some null values, we need to deal with these very strategically as we do not want to create any bias with our model and/or assume any patterns.

There seems to be a fairly equal proportion of overall scores with the recommended column. 

Firstly, I will look into filling the null overall values in correspondence to where the recommended.


This is a process which could affect our final model so we would need to be very careful.

First, we want to create two new filtered dataframes which contain the values where the recommended column's value is `yes`, and then do the same for `no` recommended

In [21]:
filt_df_yes = cleandf[cleandf['recommended'] == 'yes']
filt_df_no = cleandf[cleandf['recommended'] == 'no']

Now, we want to use this filter to calculate the respective means of the `overall` column.

In [22]:
yes_overall_mean = round(filt_df_yes['overall'].mean(),1)
no_overall_mean = round(filt_df_no['overall'].mean(),1)

print(f'Mean of yes recommended: {yes_overall_mean}')
print(f'Mean of no recommended: {no_overall_mean}')

Mean of yes recommended: 8.4
Mean of no recommended: 2.1


Now we will fill the rows that fit the condition of; where the row is null and the recommended is `yes`/`no`, we fit the respective means.

i.e. if row is null and recommended is `yes`, fill with `8.4`
if row is null and recommended is `no`, fill with `2.1`

The idea behind this approach is that, we would like to think that there is a link between the overall rating and a reason to recommend the airline, any rating between 0 and 5 is more likely to not have been recommended by a passenger, where as ratings from 6 - 10 would make sense for a positive recommendartion.

Now it is worth noting that although this seems like a sensible solution, our data is reliant on individuals submitting reviews. This method could instigate some sort of speculation around whether or not a passenger would recommend an airline. The reality is, we could never 100% be confident as we do not know how the Skytrax user was feeling towards the airline at the time of the review.

In [23]:
cleandf.loc[ (cleandf['overall'].isna()) & (cleandf['recommended']=='yes'), 'overall' ] = yes_overall_mean
cleandf.loc[ (cleandf['overall'].isna()) & (cleandf['recommended']=='no'), 'overall' ] = no_overall_mean

In [24]:
round(cleandf.isna().sum()/len(cleandf)*100, 2)

airline            0.00
overall            0.03
author             0.00
review_date        0.00
customer_review    0.00
cabin              4.01
seat_comfort       0.00
cabin_service      0.00
food_bev           0.00
entertainment      0.00
ground_service     0.00
value_for_money    0.00
recommended        2.29
dtype: float64

We are now left with just 20 null values, which make up 0.03% of our dataset. It is also worth mentioning that these are the values which do not meet the criteria of having a corresponding `recommended` value, therefore we can drop these rows as they are not useful to our model. 

Subsequently, we also have 2.29% of our `recommended` column as null values as well. We will not attempt to fill in these values as this is our target column, filling in null values would insinuate a pattern which would create bias in our final model. As a result, we would need to drop all null values as well.

In [25]:
cleandf = cleandf.dropna(axis=0)

In [26]:
#final check on final dataframe
cleandf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 63302 entries, 0 to 65932
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   airline          63302 non-null  object 
 1   overall          63302 non-null  float64
 2   author           63302 non-null  object 
 3   review_date      63302 non-null  object 
 4   customer_review  63302 non-null  object 
 5   cabin            63302 non-null  object 
 6   seat_comfort     63302 non-null  float64
 7   cabin_service    63302 non-null  float64
 8   food_bev         63302 non-null  float64
 9   entertainment    63302 non-null  float64
 10  ground_service   63302 non-null  float64
 11  value_for_money  63302 non-null  float64
 12  recommended      63302 non-null  object 
dtypes: float64(7), object(6)
memory usage: 6.8+ MB


_______

#### __1.6 Checking for duplicates__

Now we have a dataframe with 63,302 entries, with no null values. For a final check, we will quickly look into duplicated rows, and if there are any, we will drop them as we should not have a case where the same author has had the same experience twice and left the exact same review for that experience. This will reduce collinearity in our model and reduce bias.

In [27]:
print(f"Number of duplicates : {cleandf.duplicated(subset=['customer_review', 'author']).sum()}")

Number of duplicates : 4603


We have established 4,603 duplicate rows, and as mentioned earlier, this would create bias in our final model. Therefore, we must drop the duplicates.

In [28]:
cleandf = cleandf.drop_duplicates(keep='last', subset=['customer_review', 'author']).copy()

In [29]:
print(f"There are now {cleandf.duplicated(subset=['customer_review', 'author']).sum()} duplicates")

There are now 0 duplicates


In [30]:
# final dataframe check
cleandf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 58699 entries, 0 to 65932
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   airline          58699 non-null  object 
 1   overall          58699 non-null  float64
 2   author           58699 non-null  object 
 3   review_date      58699 non-null  object 
 4   customer_review  58699 non-null  object 
 5   cabin            58699 non-null  object 
 6   seat_comfort     58699 non-null  float64
 7   cabin_service    58699 non-null  float64
 8   food_bev         58699 non-null  float64
 9   entertainment    58699 non-null  float64
 10  ground_service   58699 non-null  float64
 11  value_for_money  58699 non-null  float64
 12  recommended      58699 non-null  object 
dtypes: float64(7), object(6)
memory usage: 6.3+ MB


We now have a final dataframe which has been cleaned, with all null values dealt with appropriately and duplicates removed. This dataframe is now ready for modelling.

Before we get into modelling, we will sort out the data types.

#### __1.7 Correcting Data Types__

The main data type I can notice is the fact of our `review_date` column being an object, I will proceed to change this to `datetime` as it will ensure not only clean data, but the ability to plot visualisations to describe changes overtime.

In [31]:
#select a sample of the data from the column of interest to identify the format
cleandf['review_date'].sample(10)

38903       5th August 2013
37537    15th November 2015
29627       20th March 2015
10996    20th December 2016
60934       16th April 2013
7363     19th December 2011
31852    14th November 2012
57388     13th January 2016
62093         13th May 2015
4958        29th April 2018
Name: review_date, dtype: object

In [32]:
#using python to make the conversion to datetime
from dateutil.parser import parse
cleandf['review_date'] = cleandf['review_date'].apply(lambda x: parse(x, fuzzy=True))

In [33]:
#checking if changes have occured
cleandf.head()

airline  overall               author review_date  \
0  Turkish Airlines      7.0  Christopher Hackley  2019-05-08   
1  Turkish Airlines      2.0        Adriana Pisoi  2019-05-07   
2  Turkish Airlines      3.0            M Galerko  2019-05-07   
3  Turkish Airlines     10.0          Zeshan Shah  2019-05-06   
4  Turkish Airlines      1.0           Pooja Jain  2019-05-06   

                                     customer_review          cabin  \
0  london to izmir via istanbul  first time iâ  d...  Economy Class   
1  stanbul to bucharest  we make our check in in ...  Economy Class   
2  ome to prishtina via istanbul  i flew with thi...  Economy Class   
3  lew on turkish airlines iad ist khi and return...  Economy Class   
4  mumbai to dublin via istanbul  never book turk...  Economy Class   

   seat_comfort  cabin_service  food_bev  entertainment  ground_service  \
0           4.0            5.0       4.0            4.0             2.0   
1           4.0            1.0       1.0            1.0             1.0   
2           1.0            4.0       1.0            3.0             1.0   
3           4.0            5.0       5.0            5.0             5.0   
4           1.0            1.0       1.0            1.0             1.0   

   value_for_money recommended  
0              4.0         yes  
1              1.0          no  
2              2.0          no  
3              5.0         yes  
4              1.0          no

This looks much cleaner! Zero null values and duplicated rows, the right data types. this dataset is ready for EDA.

In [34]:
#final check to ensure data is ready.
cleandf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 58699 entries, 0 to 65932
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   airline          58699 non-null  object        
 1   overall          58699 non-null  float64       
 2   author           58699 non-null  object        
 3   review_date      58699 non-null  datetime64[ns]
 4   customer_review  58699 non-null  object        
 5   cabin            58699 non-null  object        
 6   seat_comfort     58699 non-null  float64       
 7   cabin_service    58699 non-null  float64       
 8   food_bev         58699 non-null  float64       
 9   entertainment    58699 non-null  float64       
 10  ground_service   58699 non-null  float64       
 11  value_for_money  58699 non-null  float64       
 12  recommended      58699 non-null  object        
dtypes: datetime64[ns](1), float64(7), object(5)
memory usage: 6.3+ MB


In [35]:
#collect a sample of the data for better viewability
cleandf.sample(3).T

63853  \
airline                                                  Norwegian   
overall                                                        2.0   
author                                                 Sally Munns   
review_date                                    2016-06-28 00:00:00   
customer_review  he outward and return trip from london gatwick...   
cabin                                                Economy Class   
seat_comfort                                                   3.0   
cabin_service                                                  1.0   
food_bev                                                   2.90817   
entertainment                                             2.863372   
ground_service                                                 4.0   
value_for_money                                                4.0   
recommended                                                     no   

                                                             52630  \
airline                                                   Wizz Air   
overall                                                        1.0   
author                                                   P Dalenca   
review_date                                    2018-04-27 00:00:00   
customer_review  luton to timisoara  horrible experience  grump...   
cabin                                                Economy Class   
seat_comfort                                                   1.0   
cabin_service                                                  2.0   
food_bev                                                       1.0   
entertainment                                                  1.0   
ground_service                                                 1.0   
value_for_money                                                2.0   
recommended                                                     no   

                                                             15641  
airline                                            Jetblue Airways  
overall                                                        1.0  
author                                                Julie Graham  
review_date                                    2018-11-26 00:00:00  
customer_review  hartford to tampa  this is the most unprofessi...  
cabin                                                Economy Class  
seat_comfort                                                   2.0  
cabin_service                                                  2.0  
food_bev                                                       1.0  
entertainment                                             2.863372  
ground_service                                                 1.0  
value_for_money                                                2.0  
recommended                                                     no

#### __1.8 Creating Features__

Create a `Year` and `Month` column

I now want to extract the year and month from the `review_date` column so it can be treated as numerical data for the purpose of modelling and precise analysis.

I will only pull the years and months and not the days as it will be too specific to plot and might cause confusion. 

The reason for this is that I would be able to identify trend patterns over time which cann also help pin point anomalies i.e. where there was an unexpected dip in the number of positive reviews.

In [36]:
# Extracting year and month from date and creating new columns
cleandf['review_year'], cleandf['review_month'] = cleandf['review_date'].dt.year, cleandf['review_date'].dt.month

In [37]:
# Dropping original release_date column to avoid multicolliniearity 
cleandf = cleandf.drop(columns=['review_date'])

In [38]:
# View new columns
cleandf[['review_year','review_month']].head()

review_year  review_month
0         2019             5
1         2019             5
2         2019             5
3         2019             5
4         2019             5

_____

#### __1.9 Saving Work__


As we can see, the dataframe now looks good and ready for analysis. I will now save this work into a .pkl file and continue to my exploratory data analysis.

In [39]:
cleandf.columns

Index(['airline', 'overall', 'author', 'customer_review', 'cabin',
       'seat_comfort', 'cabin_service', 'food_bev', 'entertainment',
       'ground_service', 'value_for_money', 'recommended', 'review_year',
       'review_month'],
      dtype='object')

In [40]:
#save data as .pkl file in the data folder, following the same file directory.
joblib.dump(cleandf, '../data/cleandf.pkl')

['../data/cleandf.pkl']

**Notebook Ending Summary**

In this notebook, I have completed some data cleaning to ensure the data is of its highest quality before analysis and processing may occur. This will give the most accurate analysis, and thus more accurate findings.

In the next notebook, I will be exploring my cleaned data by doing eploratory data analysis.